### DISCUSSION TITLE
Raucb files integrity validation issue

### DISCUSSION THREAD

**Discussion by picostef** [rauc/rauc#1176] [Q&A] [answered]
Hello all, 
I am using rauc 1.8.0 within a custom yocto distribution for iMX8, and I am a bit puzzled with the raucb file integrity. To me the files were digitally signed, which would guarantee me some protection over file modification (over network transfer for example). 
However, I just witnessed a raucb file passing the validation of the file but being corrupted at copy (which of course crashed my filesystem :) )
I verified the md5sum on the source and on the target and they were different. I wonder how it is possible or what I have misunderstood.
How is it possible to start an update with a corrupted file?
Thanks for your help !
Stéphane

PS: Here are some details:
**On source:**

```
$ md5sum image.raucb
925d0bf870f0cb48a317b693fd0a44b1  image.raucb

$ ./poky-tmp/deploy/tools/rauc --keyring ./meta-hello/recipes-core/rauc/rauc/ca.cert.pem   --cert ./meta-hello/files/development-1.cert.pem info ~/tmp/image.raucb
rauc-Message: 13:14:51.529: Reading bundle: /home/sources/tmp/UPDATE/image.raucb
rauc-Message: 13:14:51.530: Verifying bundle signature...
rauc-Message: 13:14:51.530: Verified inline signature by 'O = Test Org, CN = Test Org Development-1'
Compatible:     'XXXX'
Version:        'Y.Y.Y'
Description:    ''
Build:          '20230626123747'
Hooks:          ''
Bundle Format:  verity
  Verity Salt:  '751fbf2a5a35fd07c0602c05300ead303a8eef93fee9a00284c66111103f06d0'
  Verity Hash:  '260012e0beaaa0ec43246e98dafa7f67d08d6513e3b3ce5e6fdd039ba0f7c22e'
  Verity Size:  1126400

1 Image:
  [factory]
        Filename:  factory.tar.gz
        Checksum:  6e9389e1267697965f93870da5a5b8f17894af7f02f3b3a30bde924c1ea687bd
        Size:      143153868
        Hooks:

Certificate Chain:
 0 Subject: O = Test Org, CN = Test Org Development-1
   Issuer: O = Test Org, CN = Test Org rauc CA Development
   SPKI sha256: 3F:E0:E0:73:27:DB:07:3E:69:F4:07:86:B4:B7:26:73:41:B8:2A:1B:3C:72:D2:6A:3B:D7:52:AD:CB:9E:E8:6E
   Not Before: Jan  1 00:00:00 1970 GMT
   Not After:  Dec 31 23:59:59 9999 GMT
 1 Subject: O = Test Org, CN = Test Org rauc CA Development
   Issuer: O = Test Org, CN = Test Org rauc CA Development
   SPKI sha256: DA:9D:78:5E:54:FA:B0:53:F0:4C:C5:F6:C6:E3:7F:7B:78:25:91:99:A2:B1:6D:80:0C:A6:6D:98:4D:B5:51:6D
   Not Before: Jan  1 00:00:00 1970 GMT
   Not After:  Dec 31 23:59:59 9999 GMT

$
```
**On target:**
```
target# md5sum image.raucb
ec8325fb4ce5b62c33300d9a5d0f9f24  image.raucb
target# rauc info image.raucb
rauc-Message: 12:33:46.150: Using central status file /rauc-data/central-status.raucs
rauc-Message: 12:33:46.155: valid /etc/rauc/system.conf found, using it
rauc-Message: 12:33:46.155: Reading bundle: /mnt/update/image.raucb
rauc-Message: 12:33:46.158: Verifying bundle signature...
rauc-Message: 12:33:46.161: Verified inline signature by 'O = Test Org, CN = Test Org Development-1'
Compatible:     'XXXX'
Version:        'Y.Y.Y'
Description:    ''
Build:          '20230626123747'
Bundle Format:  verity
  Verity Salt:  '751fbf2a5a35fd07c0602c05300ead303a8eef93fee9a00284c66111103f06d0'
  Verity Hash:  '260012e0beaaa0ec43246e98dafa7f67d08d6513e3b3ce5e6fdd039ba0f7c22e'
  Verity Size:  1126400

1 Image:
  [factory]
        Filename:  factory.tar.gz
        Checksum:  6e9389e1267697965f93870da5a5b8f17894af7f02f3b3a30bde924c1ea687bd
        Size:      143153868
        Hooks:

Certificate Chain:
 0 Subject: O = Test Org, CN = Test Org Development-1
   Issuer: O = Test Org, CN = Test Org rauc CA Development
   SPKI sha256: 3F:E0:E0:73:27:DB:07:3E:69:F4:07:86:B4:B7:26:73:41:B8:2A:1B:3C:72:D2:6A:3B:D7:52:AD:CB:9E:E8:6E
   Not Before: Jan  1 00:00:00 1970 GMT
   Not After:  Dec 31 23:59:59 9999 GMT
 1 Subject: O = Test Org, CN = Test Org rauc CA Development
   Issuer: O = Test Org, CN = Test Org rauc CA Development
   SPKI sha256: DA:9D:78:5E:54:FA:B0:53:F0:4C:C5:F6:C6:E3:7F:7B:78:25:91:99:A2:B1:6D:80:0C:A6:6D:98:4D:B5:51:6D
   Not Before: Jan  1 00:00:00 1970 GMT
   Not After:  Dec 31 23:59:59 9999 GMT

target# rauc install image.raucb -d
rauc-Message: 12:40:23.579: Debug log domains: 'rauc'
(rauc:706): rauc-DEBUG: 12:40:23.587: install started
(rauc:706): rauc-DEBUG: 12:40:23.587: input bundle: /mnt/update/image.raucb
(rauc:706): rauc-DEBUG: 12:40:23.596: Trying to contact rauc service
installing
  0% Installing
  0% Determining slot states
 20% Determining slot states done.
 20% Checking bundle
 20% Verifying signature
 40% Verifying signature done.
 40% Checking bundle done.
 40% Checking manifest contents
 60% Checking manifest contents done.
 60% Determining target install group
 80% Determining target install group done.
 80% Updating slots
 80% Checking slot factory.0
 90% Checking slot factory.0 done.
 90% Copying image to factory.0
[ 2246.097616] device-mapper: verity: 7:1: data block 7359 is corrupted
[ 2246.106163] SQUASHFS error: Failed to read block 0x1cae14f: -5
[ 2246.112061] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.118541] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.125610] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.132064] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.139149] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.145603] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.152657] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.159110] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.166164] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.172616] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.179669] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.186122] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.193177] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.199630] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.206686] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.213139] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.220193] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.226647] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.233724] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.240176] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.247231] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.253683] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.260738] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.267190] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.274245] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.280697] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.287752] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.294203] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.301258] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.307710] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.314765] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.321217] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.328272] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.334723] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.341778] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.348229] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.355284] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.361761] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.368817] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.375269] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.382325] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.388777] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.395943] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.402398] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.409541] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.415995] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.423050] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.429503] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.436559] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.443010] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.450064] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.456516] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.463571] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.470023] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.477077] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.483529] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.490596] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.497048] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.504103] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.510555] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.517609] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.524061] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.531116] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.537568] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.555783] device-mapper: verity: 7:1: data block 7359 is corrupted
[ 2246.563269] SQUASHFS error: Failed to read block 0x1cae14f: -5
[ 2246.569156] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.575633] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
[ 2246.583009] SQUASHFS error: Unable to read data cache entry [1cae14f]
[ 2246.589464] SQUASHFS error: Unable to read page, block 1cae14f, size 1020000
 99% Copying image to factory.0 failed.
 99% Updating slots failed.
100% Installing failed.
LastError: Installation error: Failed updating slot factory.0: failed to run tar extract: Child process exited with code 1
Installing `/mnt/update/image.raucb` failed
```

**Comment by jluebbe** [comment, marked as answer]
RAUC uses dm-verity to verify each block of the bundle payload as it is read. This allows us to support streaming and reduce the installation time as verification is not a serialized with installation. (see https://rauc.readthedocs.io/en/latest/reference.html#verity-format)

`[ 2246.097616] device-mapper: verity: 7:1: data block 7359 is corrupted` is dm-verity detecting the modification. Instead of the corrupted data, dm-verity reports an I/O error to the filesystem: `[ 2246.106163] SQUASHFS error: Failed to read block 0x1cae14f: -5` (5 is EIO). squashfs will the handle and pass along the I/O error to userspace.

At this point, the installation is aborted. As the installation only modifies inactive slots, there is no "damage" to the active system (similar to how RAUC is robust against power loss). Then you can retry the installation with an uncorrupted bundle.

Version 1.10 introduced the `pre-check` mode, which will read the full bundle payload before starting the installation, but this is only needed for special cases where you want to update multiple systems consistently from a single bundle.

**Reply by airtower-luna** [reply]
> Version 1.10 introduced the pre-check mode, which will read the full bundle payload before starting the installation, but this is only needed for special cases where you want to update multiple systems consistently from a single bundle.

Is there some way to run such a check outside the install command? My use case is background download of upgrade bundles (as in, periodically check for available upgrades, download the bundle if there is one, and ask the user to start installation when ready). I think it'd make sense to verify the whole bundle after download, so a corrupted download can be discarded and retried automatically (probably with some delay) without annoying the user with a failed install attempt first.

**Reply by jluebbe** [reply]
There's currently no way to run checks. Also, with streaming, this wouldn't really be useful.

An alternative would be to use adaptive updates and install the bundle without activating. Then, when the user actually wants to install the update, run the installation again and activate it (see `activate-installed` in https://rauc.readthedocs.io/en/latest/reference.html#system-configuration-file). This second installation would be very fast as it only needs to check the target slot contents without actually writing any data.

If something like that would work for you, perhaps and explicit "preload" mode would be useful, as `activate-installed` can only be set statically in the `system.conf`.

**Reply by airtower-luna** [reply]
Thank you, that looks promising! If it works, using that plus streaming would also have the advantage that I wouldn't need to make sure there's room on a data partition to hold the bundle, and not writing twice is nice for flash too.

> If something like that would work for you, perhaps and explicit "preload" mode would be useful, as activate-installed can only be set statically in the system.conf.

I should be able to set it dynamically using `--confopt`, right? My idea would be to automatically run an install with `--confopt=system:activate-installed=false`, and then either run another install without it, or do `rauc status mark-active other` when the user wants to update. Not that a shortcut wouldn't be convenient. :slightly_smiling_face:

**Reply by jluebbe** [reply]
`--confopt` would only help if you're not using the service. `rauc status mark-active other` would be the current way without restarting the service.

**Reply by jluebbe** [reply]
I've added a RFE issue: https://github.com/rauc/rauc/issues/1718

Note that slots with type `boot-*` are switched automatically.

**Reply by airtower-luna** [reply]
I'm not using the service, so that part works. The one issue I'm still having is that with this approach handlers and hooks still run, which they shouldn't until the user actually chooses to update. I assume I could disable a handler configured in `system.conf` with `--confopt`, too, but is there a way to skip hooks/handlers defined in the bundle?

**Comment by picostef** [comment]
Thanks for your answer but then I clearly misunderstood part of the documentation. :/
My configuration is an asymetric A/B configuration so I update the single rootfs slot from a recovery partition which updates the single applicative rootfs. I supposed that the signature of teh raucb was checking the integrity of the full archive so that we can start the update.
With what you described I guess I nedde to validate first the integrity of the archive (mds5sum would probably be enough).
Do I need to integrate this as an install hook from the bundle or is it not reliable enough and I should externalize the process of checking the MD5Sum outside of rauc?
Thanks for you help!
Stéphane

**Comment by jluebbe** [comment]
> Thanks for your answer but then I clearly misunderstood part of the documentation. :/ My configuration is an asymmetric A/B configuration so I update the single rootfs slot from a recovery partition which updates the single applicative rootfs.
> I supposed that the signature of the raucb was checking the integrity of the full archive so that we can start the update. With what you described I guess I need to validate first the integrity of the archive (md5sum would probably be enough).

RAUC (via dm-verity) is checking the integrity of all *parts* of the bundle which are actually *used* during installation. So there is no need for additional validation before starting the installation.

> Do I need to integrate this as an install hook from the bundle or is it not reliable enough and I should externalize the process of checking the MD5Sum outside of rauc?

RAUC always performs full cryptographic authentication of data read from the bundle.

The idea is as follows: Do be robust against errors (power loss, crash, ...) during installation, RAUC never touches the slots it is running from. So, an installation can always be retried on error. A bundle corruption detected during installation is just one kind of error that might happen during installation, but it doesn't need to be handled differently. Installation it can always be retried.

**Reply by jluebbe** [reply]
Trying to clarify: even if you check for bundle corruption before starting the installation, it may still fail due to other causes. Handling all these causes the same way reduces complexity. As RAUC deactivates the target slot before starting the installation, that slot will never be selected for booting unless the installation is successful (and it is reactivated).

**Reply by picostef** [reply]
I understand now how this works.
However, in my use case I need to be sure the file is correct before to start the update. I work with a single rootfs slot and a recovery slot. So I can retry the update as long as the original raucb archive is correct (the file is downloaded and to limit the size of the recovery slot, I removed the app allowing to download the raucb). 
So in my case I need to validate that the raucb archve was actually correctly downloaded (outside or rauc, now I get it). On top of it, you are right, I need also to be sure that revocery slot is able to access a valid rootfs to recover inany circumstances. Unfortunately a secure full A/B update is not possible due to size limit.
Thanks for all these clarifications !
